libp2p/muxers/yamux/yamux.nim

```diff
@@ -0,0 +1,485 @@
+      if isLocal:
+        try: await channel.conn.write(YamuxHeader.data(channel.id, 0, {Rst}))
+        except LPStreamEOFError as exc: discard
+      await channel.close()
+    if not channel.closedRemotely.done():
+      await channel.remoteClosed()
+    channel.receivedData.fire()
+  if not isLocal:
+    # If we reset locally, we want to flush up to a maximum of recvWindow
+    # bytes. We use the recvWindow in the proc cleanupChann.
+    channel.recvWindow = 0
+
+proc updateRecvWindow(channel: YamuxChannel) {.async.} =
+  let inWindow = channel.recvWindow + channel.recvQueue.len
+  if inWindow > channel.maxRecvWindow div 2:
+    return
+
+  let delta = channel.maxRecvWindow - inWindow
+  channel.recvWindow.inc(delta)
+  await channel.conn.write(YamuxHeader.windowUpdate(
+    channel.id,
+    delta.uint32
+  ))
+  trace "increasing the recvWindow", delta
+
+method readOnce*(
+  channel: YamuxChannel,
+  pbytes: pointer,
+  nbytes: int):
+  Future[int] {.async.} =
+
+  if channel.returnedEof: raise newLPStreamEOFError()
+  if channel.recvQueue.len == 0:
+    channel.receivedData.clear()
+    await channel.closedRemotely or channel.receivedData.wait()
+    if channel.closedRemotely.done() and channel.recvQueue.len == 0:
+      channel.returnedEof = true
+      return 0
+
+  let toRead = min(channel.recvQueue.len, nbytes)
+
+  var p = cast[ptr UncheckedArray[byte]](pbytes)
+  toOpenArray(p, 0, nbytes - 1)[0..<toRead] = channel.recvQueue.toOpenArray(0, toRead - 1)
+  channel.recvQueue = channel.recvQueue[toRead..^1]
+
+  # We made some room in the recv buffer let the peer know
+  await channel.updateRecvWindow()
+  channel.activity = true
+  return toRead
+
+proc gotDataFromRemote(channel: YamuxChannel, b: seq[byte]) {.async.} =
+  channel.recvWindow -= b.len
+  channel.recvQueue = channel.recvQueue.concat(b)
+  channel.receivedData.fire()
+  await channel.updateRecvWindow()
+
+proc setMaxRecvWindow*(channel: YamuxChannel, maxRecvWindow: int) =
+  channel.maxRecvWindow = maxRecvWindow
+
+proc setupHeader(channel: YamuxChannel, header: var YamuxHeader) =
+  if not channel.opened:
+    header.flags.incl(
+      if channel.isSrc:
+        MsgFlags.Syn
+      else:
+        MsgFlags.Ack
+    )
+    channel.opened = true
+
+proc trySend(channel: YamuxChannel) {.async.} =
+  if not channel.isSending:
+    channel.isSending = true
+    while channel.sendQueue.len != 0:
+      var futures: seq[Future[void]]
+      channel.sendQueue.keepItIf(not (it.fut.cancelled() and it.sent == 0))
+      if channel.sendWindow == 0:
+        trace "send window empty"
+        if channel.sendQueueBytes(true) > channel.maxRecvWindow:
+          await channel.reset(true)
+        break
+      if channel.sendQueue[0].data.len == 0:
+        var header = YamuxHeader.data(channel.id, 0)
+        channel.setupHeader(header)
+        let sendBuffer = toSeq(header.encode())
+        trace "send empty buffer", h = $header
+        try: await channel.conn.write(sendBuffer)
```

Comment:
You really don't like my `channel.conn.write(header)` :smile:

Reply:
I tend to forget that this proc exists